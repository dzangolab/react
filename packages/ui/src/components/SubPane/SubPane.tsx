import React, { ReactNode } from "react";
import "./subpane.css";

type Properties = {
  children: ReactNode;
  title: string;
  icon?: string;
};

const SubPane: React.FC<Properties> = ({ children }) => {
  return <div className="sub-pane">{children}</div>;
};

export default SubPane;
