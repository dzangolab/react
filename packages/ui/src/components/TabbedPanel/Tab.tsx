import React from "react";

type Properties = {
  title: string;
  icon: string;
  index: number;
  isActive: boolean;
  handleClick: (index: number) => void;
};

const Tab: React.FC<Properties> = ({
  title,
  icon,
  index,
  isActive,
  handleClick,
}) => {
  return (
    <li
      className={`tab ${isActive ? "active" : ""}`}
      onClick={() => handleClick(index)}
    >
      {icon ? <img src={icon} className="tab-icon" alt={title} /> : null}
      {title}
    </li>
  );
};

export default Tab;
