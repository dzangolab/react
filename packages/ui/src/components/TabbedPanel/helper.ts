import type { TKeymap, TOrientation, TPosition } from "./types";
import type { KeyboardEvent } from "react";

const getOrientation = (position: TPosition) => {
  let orientation: TOrientation;

  switch (position) {
    case "top":
    case "bottom":
      orientation = "horizontal";
      break;

    case "left":
    case "right":
      orientation = "vertical";
      break;

    default:
      orientation = "horizontal";
      break;
  }

  return orientation;
};

const onTabDown = (
  active: number,
  event: KeyboardEvent<HTMLButtonElement>,
  count: number,
  handleIndex: (value: number) => void,
  orientation: TOrientation
) => {
  const nextTab = () => handleIndex((active + 1) % count);
  const previousTab = () => handleIndex((active - 1 + count) % count);
  const firstTab = () => handleIndex(0);
  const lastTab = () => handleIndex(count - 1);
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  const doNothing = () => {};

  // https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code
  const keyMap: TKeymap = {
    ArrowRight: orientation === "horizontal" ? nextTab : doNothing,
    ArrowDown: orientation === "vertical" ? nextTab : doNothing,
    ArrowLeft: orientation === "horizontal" ? previousTab : doNothing,
    ArrowUp: orientation === "vertical" ? previousTab : doNothing,
    Home: firstTab,
    End: lastTab,
  };

  const action = keyMap[event.key];
  if (action) {
    event.preventDefault();
    action();
  }
};

export { getOrientation, onTabDown };
