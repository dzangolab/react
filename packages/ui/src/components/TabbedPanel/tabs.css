.tab {
  color: #444;
}

.tab button {
  align-items: center;
  background: var(--tab-background, #eee);
  border-bottom: 0;
  border: 0;
  border: var(--tab-border, 1px solid #ddd);
  cursor: pointer;
  display: flex;
  outline: 0;
  padding: 15px 25px;
  transition: background-color 0.25s ease-in-out;
}

.tab.active button {
  background: var(--tab-background-active, #fff);
}

.tabbed-panel {
  display: flex;
}

.tab button:focus-visible {
  outline-offset: -1px;
  outline: var(--tab-focus-outline, 1px solid #1c4786);
}

.tab:not(.active) button:hover {
  background: var(--tab-background-hover, #ddd);
}

.tab-icon {
  height: 16px;
  margin-right: 5px;
  object-fit: contain;
  width: 16px;
}

.tab-list {
  display: flex;
  flex-direction: column;
  list-style: none;
  margin: 0 0 10px;
  padding: 0;
}

.tab-list button {
  width: 100%;
  justify-content: center;
}

.tabbed-panel {
  display: flex;
  width: 100%;
  flex-direction: column;
}

.tabbed-panel.bottom {
  flex-direction: column-reverse;
}

.tab span {
  font-size: 1rem;
  width: max-content;
}

.tab {
  width: max-content;
}

.tabbed-panel:not(.right) .tab-list .tab:not(:last-child) button {
  border-right-width: 0;
}

.tabbed-panel.right {
  flex-direction: row-reverse;
}

.tabbed-panel.right .sub-pane {
  flex-grow: 1;
}

.tabbed-panel.right .tab-list {
  align-items: flex-end;
}

.tabbed-panel.right .tab,
.tabbed-panel.left .tab {
  width: 100%;
}

.tab-list button {
  justify-content: flex-start;
}

.tabbed-panel.right .tab-list button {
  justify-content: flex-end;
}

.tabbed-panel.left {
  flex-direction: row;
}

.tabbed-panel.left .tab-list {
  align-items: flex-start;
}

.tabbed-panel.top .tab-list,
.tabbed-panel.bottom .tab-list {
  flex-direction: row;
}

.tabbed-panel.left .tab.active button {
  border-right: var(--bottom-transparent, 1px solid transparent);
}

.tabbed-panel.top .tab.active button {
  border-bottom: var(--top-transparent, 1px solid transparent);
}

.tabbed-panel.right .tab.active button {
  border-left: var(--bottom-transparent, 1px solid transparent);
}

.tabbed-panel.bottom .tab.active button {
  border-top: var(--bottom-transparent, 1px solid transparent);
}
