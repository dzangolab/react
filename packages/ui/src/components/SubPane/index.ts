import SubPane from "./SubPane";

export { SubPane };
