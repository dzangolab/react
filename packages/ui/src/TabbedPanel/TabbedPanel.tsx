import React, { useEffect, useMemo, useRef, useState } from "react";

import { getOrientation, onTabDown } from "./helper";

import type { Properties } from "./types";

const TabbedPanel: React.FC<Properties> = ({
  children,
  defaultActiveIndex = 0,
  position = "top",
  id = "",
  persistState = false,
}) => {
  const [active, setActive] = useState<number | null>(null);
  const tabReferences = useRef<(HTMLButtonElement | null)[]>([]);
  const childNodes = Array.isArray(children) ? children : [children];

  const activeTabStorageKey = useMemo(() => `tab-${id}-active-index`, [id]);

  useEffect(() => {
    if (persistState && id) {
      const activeTabIndex = localStorage.getItem(activeTabStorageKey);

      if (activeTabIndex !== null) {
        setActive(Number(activeTabIndex));
      } else {
        setActive(defaultActiveIndex);
      }
    } else {
      setActive(defaultActiveIndex);
    }
  }, [id, persistState, defaultActiveIndex]);

  useEffect(() => {
    if (persistState && id) {
      localStorage.setItem(activeTabStorageKey, String(active));
    }
  }, [active, id, persistState]);

  const handleFocus = (index: number) => {
    const tab = tabReferences.current[index];
    if (tab) {
      tab.focus();
    }
  };

  if (!children) {
    throw new Error("TabbedPanel needs at least one children");
  }

  if (active === null) {
    return null;
  }

  return (
    <div className={`tabbed-panel ${position}`}>
      <div role="tablist" aria-orientation={getOrientation(position)}>
        {childNodes.map((item, index) => {
          const isActive = active === index;
          const title = item.props.title;
          const icon = item.props.icon;
          const key = id ? `${id}-${index}` : index;

          return (
            <button
              onKeyDown={(event) => {
                onTabDown(
                  active,
                  event,
                  childNodes.length,
                  handleFocus,
                  getOrientation(position),
                );
              }}
              onFocus={() => setActive(index)}
              ref={(element) => (tabReferences.current[index] = element)}
              onClick={() => setActive(index)}
              key={key}
              role="tab"
              aria-label={title}
              aria-disabled={isActive}
              aria-selected={isActive}
              tabIndex={isActive ? 0 : -1}
              className={isActive ? "active" : ""}
            >
              {icon ? (
                <img src={icon} alt="title icon" aria-hidden="true" />
              ) : null}
              <span>{title}</span>
            </button>
          );
        })}
      </div>
      <div role="tabpanel">{childNodes[active]}</div>
    </div>
  );
};

export default TabbedPanel;
