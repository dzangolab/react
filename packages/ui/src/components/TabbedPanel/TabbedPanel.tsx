import React, { useId, useState } from "react";

import { getOrientation, onTabDown } from "./helper";

import "./tabs.css";

import type { Properties } from "./types";

const TabbedPanel: React.FC<Properties> = ({ children, position = "top" }) => {
  const id = useId();
  const [active, setActive] = useState(0);
  const childNodes = Array.isArray(children) ? children : [children];

  const handleIndex = (index: number) => {
    setActive(index);
  };

  if (!children) {
    throw new Error("TabbedPanel needs at least one children");
  }

  return (
    <div className={`tabbed-panel ${position}`}>
      <div
        aria-label="tabs"
        role="tablist"
        aria-orientation={getOrientation(position)}
      >
        {childNodes.map((item, index) => (
          <button
            onKeyDown={(event) => {
              onTabDown(
                active,
                event,
                childNodes.length,
                handleIndex,
                getOrientation(position)
              );
            }}
            onClick={() => handleIndex(index)}
            key={`${id}-${index}`}
            role="tab"
            aria-selected={active === index}
            tabIndex={index === active ? 0 : -1}
            className={active === index ? "active" : ""}
          >
            {item.props.icon ? (
              <img src={item.props.icon} alt="title icon" />
            ) : null}
            <span>{item.props.title}</span>
          </button>
        ))}
      </div>
      <section role="tabpanel" tabIndex={0}>
        {childNodes[active]}
      </section>
    </div>
  );
};

export default TabbedPanel;
