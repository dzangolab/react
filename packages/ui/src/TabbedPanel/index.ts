import TabbedPanel from "./TabbedPanel";
import { clearSavedTabState } from "./utils";

export { TabbedPanel, clearSavedTabState };
