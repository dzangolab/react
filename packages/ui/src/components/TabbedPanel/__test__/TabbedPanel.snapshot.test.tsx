import React from "react";
import renderer, { create } from "react-test-renderer";
import { expect, test } from "vitest";

import { SubPane } from "../../SubPane";
import TabbedPanel from "../TabbedPanel";

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();

  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);

  return result as renderer.ReactTestRendererJSON;
}

test("Component matches snapshot", () => {
  const component = create(
    <TabbedPanel>
      <SubPane icon="icon1.jpg" title="Topic 1">
        Pane 1 content
      </SubPane>
      <SubPane icon="icon2.jpg" title="Topic 2">
        Pane 2 content
      </SubPane>
    </TabbedPanel>
  );
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
