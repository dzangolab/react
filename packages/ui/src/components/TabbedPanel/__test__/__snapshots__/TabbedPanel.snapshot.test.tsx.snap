// Vitest Snapshot v1

exports[`Component matches snapshot 1`] = `
<div
  className="tabbed-panel top"
>
  <div
    aria-orientation="horizontal"
    role="tablist"
  >
    <button
      aria-disabled={true}
      aria-label="Topic 1"
      aria-selected={true}
      className="active"
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      role="tab"
      tabIndex={0}
    >
      <img
        alt="title icon"
        aria-hidden="true"
        src="icon1.jpg"
      />
      <span>
        Topic 1
      </span>
    </button>
    <button
      aria-disabled={false}
      aria-label="Topic 2"
      aria-selected={false}
      className=""
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      role="tab"
      tabIndex={-1}
    >
      <img
        alt="title icon"
        aria-hidden="true"
        src="icon2.jpg"
      />
      <span>
        Topic 2
      </span>
    </button>
  </div>
  <div
    role="tabpanel"
    tabIndex={0}
  >
    <div
      className="sub-pane"
    >
      Pane 1 content
    </div>
  </div>
</div>
`;
